File: tests/runners/python/tests/08_complex_schema_test.py
from .schema import (
    GameCharacterPlayer,
    GameCharacterPlayerStatus,
    GameCharacterStats,
    GameCommonVec3,
)
from .schema_container import SchemaContainer


def make_stats() -> GameCharacterStats:
    return GameCharacterStats(
        hp=100,
        max_hp=100,
        mp=50,
        max_mp=50,
        strength=10,
        agility=8,
        intelligence=5,
        vitality=12,
    )


def make_player(player_id: int, name: str, level: int) -> GameCharacterPlayer:
    return GameCharacterPlayer(
        id=player_id,
        name=name,
        level=level,
        experience=0,
        stats=make_stats(),
        position=GameCommonVec3(x=0.0, y=0.0, z=0.0),
        status=GameCharacterPlayerStatus.Online,
        guild_id=None,
    )


def test_container_field_and_unique_validation() -> None:
    invalid_fields = SchemaContainer()
    invalid_fields.players.add_row(
        make_player(
            1,
            "A name that is definitely longer than thirty two chars",
            101,
        )
    )

    invalid_fields_result = invalid_fields.validate_all()
    assert not invalid_fields_result.is_valid()
    assert invalid_fields_result.error_count() == 2
    assert any(
        error.constraint_type == "MaxLength" and error.field_name == "name"
        for error in invalid_fields_result.errors
    )
    assert any(
        error.constraint_type == "Range" and error.field_name == "level"
        for error in invalid_fields_result.errors
    )

    duplicate = SchemaContainer()
    duplicate.players.add_row(make_player(1, "Hero A", 10))
    duplicate.players.add_row(make_player(1, "Hero B", 10))

    duplicate_result = duplicate.validate_all()
    assert not duplicate_result.is_valid()
    assert any(error.constraint_type == "Unique" for error in duplicate_result.errors)


test_container_field_and_unique_validation()

File: tests/runners/go/tests/08_complex_schema_test.go
package polygen

import "testing"

func makeComplexStats() Stats {
	return Stats{
		Hp:           100,
		MaxHp:        100,
		Mp:           50,
		MaxMp:        50,
		Strength:     10,
		Agility:      8,
		Intelligence: 5,
		Vitality:     12,
	}
}

func makeComplexPlayer(id uint32, name string, level uint16) *Player {
	return &Player{
		Id:         id,
		Name:       name,
		Level:      level,
		Experience: 0,
		Stats:      makeComplexStats(),
		Position:   Vec3{X: 0, Y: 0, Z: 0},
		Status:     StatusOnline,
	}
}

func TestComplexContainerFieldAndUniqueValidation(t *testing.T) {
	invalidFields := NewSchemaContainer()
	invalidFields.Players.AddRow(makeComplexPlayer(
		1,
		"A name that is definitely longer than thirty two chars",
		101,
	))

	result := invalidFields.ValidateAll()
	if result.IsValid() {
		t.Fatalf("expected max_length/range violations")
	}
	if result.ErrorCount() != 2 {
		t.Fatalf("expected two field validation errors, got %d: %s", result.ErrorCount(), result.String())
	}
	seenMaxLength := false
	seenRange := false
	for _, err := range result.Errors {
		if err.ConstraintType == "MaxLength" && err.FieldName == "Name" {
			seenMaxLength = true
		}
		if err.ConstraintType == "Range" && err.FieldName == "Level" {
			seenRange = true
		}
	}
	if !seenMaxLength || !seenRange {
		t.Fatalf("expected max_length and range errors, got %#v", result.Errors)
	}

	invalidRegex := NewSchemaContainer()
	invalidRegex.Players.AddRow(makeComplexPlayer(2, "Invalid!", 10))

	regexResult := invalidRegex.ValidateAll()
	if regexResult.IsValid() {
		t.Fatalf("expected regex violation")
	}
	seenRegex := false
	for _, err := range regexResult.Errors {
		if err.ConstraintType == "Regex" && err.FieldName == "Name" {
			seenRegex = true
			break
		}
	}
	if !seenRegex {
		t.Fatalf("expected regex validation error, got %#v", regexResult.Errors)
	}

	duplicate := NewSchemaContainer()
	duplicate.Players.AddRow(makeComplexPlayer(1, "Hero A", 10))
	duplicate.Players.AddRow(makeComplexPlayer(1, "Hero B", 10))

	duplicateResult := duplicate.ValidateAll()
	if duplicateResult.IsValid() {
		t.Fatalf("expected duplicate primary key to fail validation")
	}
	seenUnique := false
	for _, err := range duplicateResult.Errors {
		if err.ConstraintType == "Unique" {
			seenUnique = true
			break
		}
	}
	if !seenUnique {
		t.Fatalf("expected unique validation error, got %#v", duplicateResult.Errors)
	}
}
